### DISCUSSION TITLE
Cad Viewer ignores the "invisible" attribute

### DISCUSSION THREAD

**Discussion by melihpeker-pointr** [mozman/ezdxf#465] [Issues] [answered]
Hi, while drawing some of the blocks of some of the DXF files via cad_viewer.py, I get problematic views. For example, a desk block can be seen via Autodesk Viewer as this:

![Screenshot 2021-06-11 at 09 26 23](https://user-images.githubusercontent.com/83574547/121641886-70b9da00-ca98-11eb-980d-c0f66fd25c4f.png)

 However when I try to view the same DXF, same place, same blocks, same layers ON, I get a view like this on CAD Viewer:

![Screenshot 2021-06-11 at 09 35 07](https://user-images.githubusercontent.com/83574547/121641941-8202e680-ca98-11eb-96de-efc43e12e2cb.png)

Is there anyone encountered a similar problem? What could be the causing of the problem? This is not a single DXF issue, I got the similar problem on some of my DXF files.

Thanks!

**Comment by mozman** [comment]
Please, upload DXF file as zip-archive.

**Reply by melihpeker-pointr** [reply]
I can't upload the DXF because of the confidentiality. I think there is a problem with reading even before drawing part. I just read the dxf with `ezdxf.readfile ` and then without doing any operation saved it via  `doc.saveas`. The size of the file decreases and Autodesk Viewer cannot open the saved file and raises the invalid file error.

**Reply by chibai** [reply]
> I can't upload the DXF because of the confidentiality. I think there is a problem with reading even before drawing part. I just read the dxf with `ezdxf.readfile ` and then without doing any operation saved it via `doc.saveas`. The size of the file decreases and Autodesk Viewer cannot open the saved file and raises the invalid file error.

You can copy the block to another dxf file to reproduce the bug.

**Reply by melihpeker-pointr** [reply]
Here one of the problematic blocks:

[Desks_021.dxf.zip](https://github.com/mozman/ezdxf/files/6636789/Desks_021.dxf.zip)

There are similar problems with another blocks as well like this one:

![*U503](https://user-images.githubusercontent.com/83574547/121653330-41f63080-caa5-11eb-9e34-3e268a171251.png)
 
The DXF regarding this block is: 
[DuplicateChairs.zip](https://github.com/mozman/ezdxf/files/6636809/fadingChairs.zip)

**Reply by mozman** [reply]
The file "\*U503.dxf" - I don't know how you can use a "*" in the filename -  can not be opened by Autodesk TrueView 2022.

BricsCAD shows this:
![image](https://user-images.githubusercontent.com/4230966/121660165-51747a00-caa3-11eb-8fc1-3b4cd7daa860.png)

and the *ezdxf* viewer shows this:
![image](https://user-images.githubusercontent.com/4230966/121660009-268a2600-caa3-11eb-8d30-943ec701a3b4.png)

Opening and saving the DXF by *ezdxf* works and the saved file can be opened by BricsCAD.

The provided file is just an invalid DXF file and it is much to big to do any debugging, you have to create a smaller file which shows the issue. If you are using AutoCAD you can use the PURGE command to get rid of unused resources. 

Saving the source DXF by BricsCAD creates a DXF file which can be loaded by TrueView, but the result is the same as in BricsCAD:

![image](https://user-images.githubusercontent.com/4230966/121661863-f0e63c80-caa4-11eb-889b-4510053ec1ef.png)

**Reply by mozman** [reply]
Running BricsCAD RECOVER command prints:
Total errors found during audit 240, fixed 240

I doubt this is an AutoCAD file.

**Reply by melihpeker-pointr** [reply]
This file is not generated by AutoCAD, I create this file using the `importer`. I imported this block from the original file to an empty dxf file and saved it to show you. As I said before, although Autodesk Viewer views without any problem, when I read it with ezdxf, it becomes like this one.

**Reply by mozman** [reply]
OK, the *ezdxf* `importer` add-on is not very reliable, and you can not extract test data to verify unknown *ezdxf* issues by using *ezdxf*!

**Reply by melihpeker-pointr** [reply]
Of course, just wanted to show you how it looks like. As I said, I think there is a problem in reading part. I'm trying to figure out the problem.

**Reply by melihpeker-pointr** [reply]
@mozman I saved the desk block as a dxf file using AutoCAD 2022. Here, first picture is the view from AutoCAD, second is from the CAD Viewer. The block: 
[Desk.zip](https://github.com/mozman/ezdxf/files/6638462/Desk.zip)

![Screenshot 2021-06-11 at 15 47 12](https://user-images.githubusercontent.com/83574547/121688848-93fe7c80-cacc-11eb-9504-75e2ccfbe615.png)
![Screenshot 2021-06-11 at 15 47 16](https://user-images.githubusercontent.com/83574547/121688854-95c84000-cacc-11eb-87bd-f83265db9723.png)

**Reply by melihpeker-pointr** [reply]
BTW, when I try to edit the Desk drawing with `BEDIT` command in AutoCAD 2022, I see the following:

![Screenshot 2021-06-11 at 17 00 14](https://user-images.githubusercontent.com/83574547/121698332-92d24d00-cad6-11eb-9f88-28777a350156.png)

I think what we see here is the all attributes and hidden objects of the block and in _ezdxf_, we are drawing all attributes and hidden objects so the drawing looks like overlapping blocks, could that be the reason @mozman ?

**Comment by mozman** [comment, marked as answer]
@mbway The visibility status of the INSERT entity seems do be dependent from the invisible attribute. But this breaks some tests in 811 and I don't know why, because the "invisible" attribute is not used in the tests. Marked the failing tests by "xfail".

Current result:
![image](https://user-images.githubusercontent.com/4230966/121698918-c4dfb100-cace-11eb-8247-52cbc3789c84.png)

Fixed version:
![image](https://user-images.githubusercontent.com/4230966/121698616-729e9000-cace-11eb-854a-731bfb14df6f.png)

**Reply by mbway** [reply]
I'm afraid I don't have time to look into this right now, sorry

**Reply by mozman** [reply]
Is hopefully fixed!

**Reply by melihpeker-pointr** [reply]
Thanks for the quick fix! Also, I realised that when we explode the blocks, the exploded entities do not have the invisible attribute so while drawing exploded entities, all of the entities are drawn.

**Reply by mozman** [reply]
It is hard to say what the expected behavior should be, because I don't know if it is possible to explode invisible block references in CAD applications. I think it is also legit to explode the content of a block reference as it is.

You have two options for handling this:
1. to just examine the content, don't explode the block, use the `virtual_entities()` method instead, which returns the content without exploding the block reference
2. the `explode()`method returns an `EntityQuery` object, which contains references to all exploded entities and set the invisible attribute by yourself: `for e in entity.explode(): e.dxf.invisible=entity.dxf.invisible`

**Comment by mozman** [comment]
Collecting facts ...

In the marked area are the same INSERT entities as above but with the "invisible" attribute set to 1:

![image](https://user-images.githubusercontent.com/4230966/121710268-bb0f7b00-cad9-11eb-8860-dd768bd25fb0.png)

Fact 1: The "invisible" flag switches the INSERT off - independently from the LAYER state.
